datalad/cmd.py

```diff
@@ -98,7 +98,208 @@ def _cleanup_output(stream, std):
+    functionality. Derived classes should be used for additional
+    specializations and convenience features.
+    """
+    __slots__ = ['cwd', 'env', '_poll_period']
+
+    def __init__(self, cwd=None, env=None, poll_period=0.1):
+        """
+        Parameters
+        ----------
+        cwd : path-like, optional
+          If given, commands are executed with this path as PWD,
+          the PWD of the parent process is used otherwise.
+        env : dict, optional
+          Environment to be passed to subprocess.Popen(). If `cwd`
+          was given, 'PWD' in the environment is set to its value.
+          This must be a complete environment definition, no values
+          from the current environment will be inherited.
+        poll_period : float, optional
+          Interval at which the running process is queried for
+          output.
+        """
+        self._poll_period = poll_period
+        self.env = env.copy() if env else None
+        # stringify to support Path instances on PY35
+        self.cwd = str(cwd) if cwd is not None else None
+        if self.cwd and env is not None:
+            # if CWD was provided, we must not make it conflict with
+            # a potential PWD setting
+            self.env['PWD'] = self.cwd
+
+    def run(self, cmd, proc_stdout=None, proc_stderr=None, stdin=None):
+        """Execute a command and communicate with it.
+
+        Parameters
+        ----------
+        cmd : list
+          Sequence of program arguments. Passing a single string means
+          that it is simply the name of the program, no complex shell
+          commands are supported.
+        proc_stdout : callable, optional
+          By default no stdout is captured, but relayed to the parent
+          process's stdout. If given, all stdout is passed as a byte-string
+          to this callable, in the chunks it was received by polling the
+          processing. The callable may transform it in any way, its output
+          (byte-string) is concatenated and provided as stdout return value.
+          The helper functions 'kill_output' and 'capture_output' are
+          provided to either swallow all output (and not relay it to the
+          parent) or to capture all output and provide it as the return
+          value.
+        proc_stderr : callable, optional
+          Like proc_stdout, but for stderr.
+        stdin : byte stream, optional
+          File descriptor like, used as stdin for the process. Passed
+          verbatim to subprocess.Popen().
+
+        Returns
+        -------
+        stdout, stderr
+          Unicode string with the cumulative standard output and error
+          of the process.
+
+        Raises
+        ------
+        CommandError
+          On execution failure (non-zero exit code) this exception is
+          raised which provides the command (cmd), stdout, stderr,
+          exit code (status), and a message identifying the failed
+          command, as properties.
```

Comment:
It will raise a FileNotFoundError if the program doesn't exist.  Do you think, since we're reworking things, we should wrap that as a CommandError?

Reply:
I was torn, and ultimately decided it would be more useful if those are distinguished. It will be a rather common exception, because any invocation with just-a-string is likely to raise it. However, I am open to alternative opinions (and in either case the docs need to be adjusted).

Reply:
Yeah, I'm on the fence, and just the doc adjustment is okay with me.

Reply:
Added now.